File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/map/overlay/NOverlayInfo.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay

import com.naver.maps.map.overlay.Overlay
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.model.enum.NOverlayType

internal data class NOverlayInfo(
    val type: NOverlayType,
    val id: String,
    val method: String? = null,
) {
    fun getOverlay(overlays: Map<String, Overlay>): Overlay? = overlays[overlayMapKey]

    /* ----- toMessageable ----- */

    val overlayMapKey = listOf(type, id).joinToString(SEPARATE_STRING)

    fun toQueryString(injectMethod: String? = null): String =
        listOf(overlayMapKey, injectMethod ?: method).joinToString(SEPARATE_STRING)

    fun toMap(): Map<String, Any> = mapOf("type" to type.toString(), "id" to id)

    /* ----- fromMessageable ----- */
    companion object {
        fun fromMap(rawMap: Any): NOverlayInfo = rawMap.asMap().let { map ->
            NOverlayInfo(
                type = NOverlayType.fromString(map["type"]!!.toString()),
                id = map["id"]!!.toString(),
            )
        }

        fun fromString(query: String): NOverlayInfo = query.split(SEPARATE_STRING).let {
            NOverlayInfo(
                type = NOverlayType.fromString(it[0]),
                id = it[1],
                method = it.getOrNull(2),
            )
        }

        private const val SEPARATE_STRING = "\""

        val locationOverlayInfo = NOverlayInfo(NOverlayType.LOCATION_OVERLAY, "L")
    }
}

File: ios/Classes/model/map/overlay/NOverlayInfo.swift
import NMapsMap

struct NOverlayInfo {
    let type: NOverlayType
    let id: String
    let method: String?

    init(type: NOverlayType, id: String, method: String? = nil) {
        self.type = type
        self.id = id
        self.method = method
    }

    func getOverlay(overlays: Dictionary<String, NMFOverlay>) -> NMFOverlay? {
        overlays[overlayMapKey]
    }

    /* ----- toMessageable ----- */

    var overlayMapKey: String {
        [type.rawValue, id].joined(separator: NOverlayInfo.separateString)
    }

    func toQueryString(injectMethod: String? = nil) -> String {
        [overlayMapKey, (injectMethod ?? method!)].joined(separator: NOverlayInfo.separateString)
    }

    func toMessageable() -> Dictionary<String, Any> {
        ["type": type.rawValue, "id": id]
    }

    /* ----- fromMessageable ----- */

    static func fromMessageable(_ v: Any) -> NOverlayInfo {
        let d = asDict(v)
        return NOverlayInfo(
                type: NOverlayType(rawValue: asString(d["type"]!))!,
                id: asString(d["id"]!)
        )
    }

    static func fromString(_ query: String) -> NOverlayInfo {
        let arr = query.components(separatedBy: NOverlayInfo.separateString)
        return NOverlayInfo(
                type: NOverlayType(rawValue: asString(arr[0]))!,
                id: asString(arr[1]),
                method: (arr.count > 2) ? asString(arr[2]) : nil
        )
    }

    private static let separateString = "\"";

    static let locationOverlayInfo = NOverlayInfo(type: .locationOverlay, id: "L")
}